test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +280,595 @@ func TestAdoption(t *testing.T) {
+		if err != nil {
+			return false, err
+		}
+		if newRS.Status.ReadyReplicas != *newRS.Spec.Replicas {
+			return false, nil
+		}
+		return true, nil
 	}); err != nil {
-		t.Fatal(err)
+		t.Fatalf("failed to wait all pods to become ready for rs %s: %v", rs.Name, err)
 	}
-	close(stopCh)
 }
 
-// selectors are IMMUTABLE for all API versions except extensions/v1beta1
-func TestRSSelectorImmutability(t *testing.T) {
-	s, closeFn, clientSet := rmSimpleSetup(t)
+func TestReadyAndAvailableReplicas(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
 	defer closeFn()
-	ns := framework.CreateTestingNamespace("rs-selector-immutability", s, t)
+	ns := framework.CreateTestingNamespace("test-ready-and-available-replicas", s, t)
 	defer framework.DeleteTestingNamespace(ns, s, t)
-	rs := newRS("rs", ns.Name, 0)
-	createRSsPods(t, clientSet, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
 
-	// test to ensure extensions/v1beta1 selector is mutable
-	newSelectorLabels := map[string]string{"changed_name_extensions_v1beta1": "changed_test_extensions_v1beta1"}
-	rs.Spec.Selector.MatchLabels = newSelectorLabels
-	rs.Spec.Template.Labels = newSelectorLabels
-	replicaset, err := clientSet.ExtensionsV1beta1().ReplicaSets(ns.Name).Update(rs)
+	labelMap := map[string]string{"foo": "bar"}
+	rs := newRS("rs", ns.Name, 3)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+
+	// Make .spec.minReadySeconds long enough so that we can test the replicas are unavailable
+	rs.Spec.MinReadySeconds = 3600
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
+	rs = rss[0]
+
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	defer close(stopCh)
+	waitRSStable(t, c, rs, ns.Name)
+
+	podClient := c.Core().Pods(ns.Name)
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := podClient.List(options)
 	if err != nil {
-		t.Fatalf("failed to update extensions/v1beta1 replicaset %s: %v", replicaset.Name, err)
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
 	}
-	if !reflect.DeepEqual(replicaset.Spec.Selector.MatchLabels, newSelectorLabels) {
-		t.Errorf("selector should be changed for extensions/v1beta1, expected: %v, got: %v", newSelectorLabels, replicaset.Spec.Selector.MatchLabels)
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	if len(pods.Items) != 3 {
+		t.Fatalf("len(pods) = %v, want %v", len(pods.Items), 3)
 	}
 
-	// test to ensure apps/v1beta2 selector is immutable
-	rsV1beta2, err := clientSet.AppsV1beta2().ReplicaSets(ns.Name).Get(replicaset.Name, metav1.GetOptions{})
-	if err != nil {
-		t.Fatalf("failed to get apps/v1beta2 replicaset %s: %v", replicaset.Name, err)
+	markAllPodsRunningAndReady(t, c, pods, ns.Name, *rs.Spec.Replicas)
+	waitAllPodsReady(t, c, rs, ns.Name)
+	if rs.Status.AvailableReplicas != 0 {
```

Comment:
You need to GET `rs` rather than reusing the old value returned from `createRSsPods()`.

Reply:
ya, this is a necessary step.